volatility/framework/plugins/mac/pslist_process_groups.py

```diff
@@ -0,0 +1,89 @@
+from volatility.plugins.mac import pslist
+
+vollog = logging.getLogger(__name__)
+
+class Pslist_Process_Groups(pslist.PsList):
+    """Lists the processes present in a particular mac memory image."""
+
+    @classmethod
+    def list_tasks(cls,
+                   context: interfaces.context.ContextInterface,
+                   layer_name: str,
+                   darwin_symbols: str,
+                   filter_func: Callable[[int], bool] = lambda _: False) -> \
+            Iterable[interfaces.objects.ObjectInterface]:
+        """Lists all the tasks in the primary layer.
+
+        Args:
+            context: The context to retrieve required elements (layers, symbol tables) from
+            layer_name: The name of the layer on which to operate
+            darwin_symbols: The name of the table containing the kernel symbols
+            filter_func: A function which takes a task object and returns True if the task should be ignored/filtered
+
+        Returns:
+            The list of task objects from the `layer_name` layer's `tasks` list after filtering
+        """
+
+        kernel = contexts.Module(context, darwin_symbols, layer_name, 0)
+
+        kernel_as = context.layers[layer_name]
+
+        table_size = kernel.object_from_symbol(symbol_name = "pgrphash") 
+
+        pgrphashtbl = kernel.object_from_symbol(symbol_name = "pgrphashtbl")  
+
+        proc_array = kernel.object(object_type = "array",
+                                   offset = pgrphashtbl, 
+                                   count = table_size + 1,  
+                                   subtype = kernel.get_type("pgrphashhead"))
+
+        for proc_list in proc_array:
+            # test the validity of the current element
+            # it is expected that many won't be initialized
+            try:
+                pgrp = proc_list.lh_first
+            except exceptions.PagedInvalidAddressException:
+                continue
+
+            seen_pgrps = set()
+
+            # this walks the particular process group
+            while pgrp and pgrp.vol.offset not in seen_pgrps:
+                seen_pgrps.add(pgrp.vol.offset)
+            
+                # nothing can be done if this list pointer is invalid, so move on
+                try:
+                    p = pgrp.pg_members.lh_first
+                except exceptions.PagedInvalidAddressException:
+                    break
+
+                seen_pg = set()
+                while p and p.vol.offset not in seen_pg:
+                    seen_pg.add(p.vol.offset)
+
+                    if p.is_readable():
+                        yield p
+                    
+                    try:
+                        p = p.p_pglist.le_next 
+                    except exceptions.PagedInvalidAddressException:
+                        break
+                try:
+                    pgrp = pgrp.pg_hash.le_next
+                except exceptions.PagedInvalidAddressException:
+                    break
+
+
+
```

Comment:
5:S  I really don't want to see these (pslist, pslist_process_groups, tasks) mount up, and I'd almost prefer them as different classmethods on pslist, with parameters/requirements/flags to choose between them.  Inheritance doesn't easily allow for suitable dependency verification if a breaking, but not compilation breaking, change occurs...

Happy to hear reasons for keeping them separate plugins, but I can't see why flags wouldn't work out with a single plugin instead?

Reply:
A single plugin would likely lead to people just assuming we only have support for the one method since in vol2 we have separate process enumeration plugins in Windows, Mac, and Linux. Also, the alternate methods would not appear when people used the help option to see all the plugins, instead all they would see is pslist even though it really would have numerous ways to enumerate process. The list that pslist walks can be broken by malware without negatively affecting the system whereas the other sources will have the processes due to instability if changed.

Reply:
So, `vol.py --help` would show the pslist plugin `vol.py pslist --help` would show the various options for the plugin, which would list the various means of enumerating the values.
The fact that vol2 did something a particular way doesn't imply that it's necessarily a good way of doing it and shouldn't bind us to do things that way forever after.  I'm not sure how great the user confusion would be if we document it well (the plugin docstring could mention that it contains different ways of finding these values) given we've already got other plugins that work differently to vol2, but I think developers would find it beneficial to have different methods of the same thing grouped in the same place?
Given that they all return the same data, that no one of them works properly, but one stands out as the most likely (but vulnerable to malware) means of getting the data, I'd suggest that be the default, and then a choice option to pick another method would work pretty well?  It also saves us coming up with weird names (why is `tasks` not `pslist_tasks` for example)...

None of those reasons feel solid enough to justify not having a single plugin.  I'm still open to the concept, but also still leaning towards a single plugin as the way forward?  A single plugin would mean that any single change would require the version number to be bumped, but I don't think that's unreasonable?  If we ever needed to split them out in the future, we could do so by bumping the plugin version number.  Given that all the others inherit from the single original plugin, they all (currently) share a version number, so there's no big change there...  We could even move the class methods to non-plugin classes with their own versions which pslist could check before using them, but that feels like it might complicate things unnecessarily.  Is there something else I'm not seeing?  If it's a strongly held conviction that they should be separate I'll consider that, but overall I'm not seeing the downside?

Reply:
ok, we can try to push them all into one and see how it plays out over time. Would you made merging this one into pslist.py as you described and then I can do the others?

Reply:
Sure, I'll try and mock it up tonight.  I'll push the results directly to this branch so we can follow what's going on...  5:)